drivers/greth.c

```diff
@@ -0,0 +1,564 @@
+	.nextRxBufferSize = greth_nextRxBufferSize,
+	.pktRxFinished = greth_pktRxFinished,
+	.fillRxDesc = greth_fillRxDesc,
+	.nextTxDone = greth_nextTxDone,
+	.fillTxDesc = greth_fillTxDesc,
+	.desc_size = sizeof(greth_desc_t),
+	.pkt_buf_sz = GRETH_MAX_PKT_SZ,
+	.ring_alignment = 1024,
+	.max_tx_frag = 0xFFFF,
+};
+
+static const size_t greth_ring_sizes[] = { GRETH_RX_RING_SIZE, GRETH_TX_RING_SIZE };
+
+
+static int greth_initRings(greth_state_t *state)
+{
+	int err;
+
+	err = net_initRings(state->rings, greth_ring_sizes, sizeof(state->rings) / sizeof(*state->rings), &greth_ring_ops);
+	if (err != 0) {
+		return err;
+	}
+
+	state->mmio->TX_DESC_PTR = state->tx.phys;
+	state->mmio->RX_DESC_PTR = state->rx.phys;
+
+	return 0;
+}
+
+
+static err_t greth_netifOutput(struct netif *netif, struct pbuf *p)
+{
+	greth_state_t *state = netif->state;
+	size_t nf;
+
+	if (ETH_PAD_SIZE != 0) {
+		(void)pbuf_header(p, -ETH_PAD_SIZE); /* drop the padding word */
+	}
+
+	mutexLock(state->tx_lock);
+	nf = net_transmitPacket(&state->tx, p);
+	if (nf != 0) {
+		state->mmio->CTRL |= GRETH_CTRL_TE;
+	}
+	mutexUnlock(state->tx_lock);
+
+	return nf ? ERR_OK : ERR_BUF;
+}
+
+
+static inline uint8_t greth_getByte(uint32_t v, int i)
+{
+	return (v >> (i * 8)) & 0xFF;
+}
+
+
+static void greth_readCardMac(greth_state_t *state)
+{
+	uint8_t *mac;
+	mac = (void *)&state->netif->hwaddr;
+
+	/* MAC LSB reg*/
+	mac[5] = greth_getByte(state->mmio->MAC_LSB, 0);
+	mac[4] = greth_getByte(state->mmio->MAC_LSB, 1);
+	mac[3] = greth_getByte(state->mmio->MAC_LSB, 2);
+	mac[2] = greth_getByte(state->mmio->MAC_LSB, 3);
+
+	/* MAC MSB reg*/
+	mac[1] = greth_getByte(state->mmio->MAC_MSB, 0);
+	mac[0] = greth_getByte(state->mmio->MAC_MSB, 1);
+}
+
+
+static void greth_showCardId(greth_state_t *state)
+{
+	uint8_t *mac;
+	mac = (void *)&state->netif->hwaddr;
+	greth_printf(state, "initizalized, MAC=%02x:%02x:%02x:%02x:%02x:%02x", mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
+}
+
+
+static void greth_startGreth(greth_state_t *state)
+{
+	/* clear status */
+	state->mmio->STAT =
+			GRETH_STAT_IA | GRETH_STAT_TS |
+			GRETH_STAT_TA | GRETH_STAT_RA |
+			GRETH_STAT_TI | GRETH_STAT_RI |
+			GRETH_STAT_TE | GRETH_STAT_RE;
+
+	/* interrupt mask */
+	state->mmio->CTRL |= GRETH_CTRL_TI | GRETH_CTRL_RI;
+}
+
+
+static int greth_createResources(greth_state_t *state)
+{
+	if (mutexCreate(&state->irq_lock) < 0) {
+		return -1;
+	}
+	if (condCreate(&state->irq_cond) < 0) {
+		return -1;
+	}
+	return 0;
+}
+
+
+static void greth_cleanupResources(greth_state_t *state)
+{
+	if (state->irq_lock < (handle_t)-1) {
+		resourceDestroy(state->irq_lock);
+	}
+	if (state->irq_cond < (handle_t)-1) {
+		resourceDestroy(state->irq_cond);
+	}
+}
+
+
+static int greth_initDevice(greth_state_t *state, int irq, int mdio)
+{
+	state->mmio = physmmap(state->dev_phys_addr, 0x1000);
+	if (state->mmio == MAP_FAILED) {
+		return -ENOMEM;
+	};
+
+#ifdef GRETH_DISABLE_EDCL
+	int err = greth_disable_edcl(state);
+#endif
```

Comment:
Maybe such configuration should reside in the `board_config.h` file for the target?